src/vmm/src/gdb/target.rs

```diff
@@ -0,0 +1,746 @@
+    VCPURequestError,
+    /// No currently paused Vcpu error
+    NoPausedVCpu,
+    /// Error when setting vcpu debug flags
+    VcpuKvmError,
+    /// Server socket Error
+    ServerSocketError,
+    /// Error with creating GDB thread
+    GdbThreadError,
+}
+
+impl<E> From<Error> for TargetError<E> {
+    fn from(_value: Error) -> Self {
+        TargetError::NonFatal
+    }
+}
+
+/// The debug target for firecracker. This is used the manage the debug
+/// implementation and handle requests sent via GDB
+#[derive(Debug)]
+pub struct FirecrackerTarget {
+    vmm: Arc<Mutex<Vmm>>,
+    entry_addr: GuestAddress,
+
+    /// This event fd us used to notify us of cpu debug exits
+    pub gdb_event: Receiver<usize>,
+
+    hw_breakpoints: Vec<GuestAddress>,
+    sw_breakpoints: HashMap<<GdbArch as Arch>::Usize, [u8; 1]>,
+
+    vcpu_state: Vec<VCpuState>,
+
+    paused_vcpu: Option<Tid>,
+}
+
+fn tid_to_cpuid(tid: Tid) -> usize {
+    tid.get() - 1
+}
+
+/// Converts the inernal index of a vcpu to
+/// the Tid required by GDB
+pub fn cpuid_to_tid(cpu_id: usize) -> Tid {
+    Tid::new(get_raw_tid(cpu_id)).expect("Error translating TID")
+}
+
+/// Converts the inernal index of a vcpu to
+/// the 1 indexed value for GDB
+pub fn get_raw_tid(cpu_id: usize) -> usize {
+    cpu_id + 1
+}
+
+impl FirecrackerTarget {
+    /// Creates a new Target for GDB stub. This is used as the layer between GDB and the VMM it
+    /// will handle requests from GDB and perform the appropriate actions, while also updating GDB
+    /// with the state of the VMM / VCPU's as we hit debug events
+    pub fn new(vmm: Arc<Mutex<Vmm>>, gdb_event: Receiver<usize>, entry_addr: GuestAddress) -> Self {
+        let cpu_count = {
+            vmm.lock()
+                .expect("Exception unlocking vmm")
+                .vcpus_handles
+                .len()
+        };
+
+        let vcpu_state = (0..cpu_count)
+            .map(|_| VCpuState {
+                paused: false,
+                single_step: false,
+            })
+            .collect();
+
+        Self {
+            vmm,
+            entry_addr,
+            gdb_event,
+            // We only support 4 hw breakpoints on x86 this will need to be configurable on arm
+            hw_breakpoints: Vec::with_capacity(4),
+            sw_breakpoints: HashMap::new(),
+            vcpu_state,
+
+            paused_vcpu: Tid::new(1),
+        }
+    }
+
+    fn get_paused_vcpu(&self) -> Result<Tid, Error> {
+        match self.paused_vcpu {
+            Some(res) => Ok(res),
+            None => Err(Error::NoPausedVCpu),
+        }
+    }
+
+    /// This is used to notify the target that the provided Tid
+    /// is in a paused state
+    pub fn notify_paused_vcpu(&mut self, tid: Tid) {
+        let found = &mut self.vcpu_state[tid_to_cpuid(tid)];
+        found.paused = true;
+
+        self.paused_vcpu = Some(tid);
+    }
+
+    fn resume_execution(&mut self) -> Result<(), Error> {
```

Comment:
I've tidied up the area of code somewhat, but kept the Tid, I thought I'd see how you felt it looked after the refactoring

Reply:
I had a closer look at this. Do we actually need `VcpuState::paused` as a field? I see we use it to avoid pausing a paused vcpu, or resuming a running one, but neither of those operations are actually errors. And the check in `update_kvm_debug` can be dropped if the later `NotAllowed` pattern match is changed to return `Ok(())`. 

As for `VcpuState::single_step`, I think moving it into `VcpuHandle` makes things a bit nicer, because we no longer need the `vcpu_states` list (instead we probably want to just store `vcpu_count`, so that we don't have to lock the vmm mutex just to get the count). That way, we can at least eliminate combining data from two arrays that need to be indexed the same way.

I had a look at moving a bunch of the functions that take `tid` to instead be inherent to `VcpuHandle`, in hopes that the API could be `FirecrackerTarget::vcpu_handle(tid).translate_hva(addr)?` and so on, but sadly a function of the signature `fn (&mut FirecrackerTarget) -> &mut VcpuHandle` is not possible without something like https://github.com/rust-lang/rust/issues/117108. At least I couldn't get it to work. So I think just moving `single_step` and `paused` (if it's really needed) into `VcpuHandle` is the best we can do here.

Reply:
(well, you could solve it with a closure, but that's a horrible API):

<details>

```diff
diff --git a/src/vmm/Cargo.toml b/src/vmm/Cargo.toml
index 6a6b05c91..50f0fd40d 100644
--- a/src/vmm/Cargo.toml
+++ b/src/vmm/Cargo.toml
@@ -58,7 +58,7 @@ itertools = "0.13.0"
 proptest = { version = "1.5.0", default-features = false, features = ["std"] }
 
 [features]
-default = []
+default = ["debug"]
 tracing = ["log-instrument"]
 debug = ["gdbstub", "gdbstub_arch"]
 
diff --git a/src/vmm/src/gdb/event_loop.rs b/src/vmm/src/gdb/event_loop.rs
index 6d584dcea..f55884e09 100644
--- a/src/vmm/src/gdb/event_loop.rs
+++ b/src/vmm/src/gdb/event_loop.rs
@@ -65,15 +65,17 @@ impl run_blocking::BlockingEventLoop for GdbBlockingEventLoop {
                         .map_err(WaitForStopReasonError::Target)?;
 
                     let Some(stop_response) = stop_reason else {
-                        // If we returned None this is a break which should be handled by
-                        // the guest kernel (e.g. kernel int3 self testing) so we won't notify
-                        // GDB and instead inject this back into the guest
-                        target
-                            .inject_bp_to_guest(tid)
-                            .map_err(WaitForStopReasonError::Target)?;
-                        target
-                            .request_resume(tid)
-                            .map_err(WaitForStopReasonError::Target)?;
+                        target.with_vcpu_handle(tid, |vcpu_handle| {
+                            // If we returned None this is a break which should be handled by
+                            // the guest kernel (e.g. kernel int3 self testing) so we won't notify
+                            // GDB and instead inject this back into the guest
+                            vcpu_handle
+                                .inject_bp_to_guest(target.hw_breakpoints.active_breakpoint())
+                                .map_err(WaitForStopReasonError::Target)?;
+                            vcpu_handle
+                                .request_resume()
+                                .map_err(WaitForStopReasonError::Target)
+                        })?;
 
                         trace!("Injected BP into guest early exit");
                         continue;
@@ -102,13 +104,13 @@ impl run_blocking::BlockingEventLoop for GdbBlockingEventLoop {
         target: &mut FirecrackerTarget,
     ) -> Result<Option<MultiThreadStopReason<u64>>, <FirecrackerTarget as Target>::Error> {
         // notify the target that a ctrl-c interrupt has occurred.
-        let main_core = vcpuid_to_tid(0)?;
+        let tid = vcpuid_to_tid(0)?;
+        target.with_vcpu_handle(tid, |handle| handle.request_pause())?;
 
-        target.request_pause(main_core)?;
-        target.update_paused_vcpu(main_core);
+        target.update_paused_vcpu(tid);
 
         let exit_reason = MultiThreadStopReason::SignalWithThread {
-            tid: main_core,
+            tid,
             signal: Signal::SIGINT,
         };
         Ok(Some(exit_reason))
diff --git a/src/vmm/src/gdb/target.rs b/src/vmm/src/gdb/target.rs
index f7cf88814..cde666731 100644
--- a/src/vmm/src/gdb/target.rs
+++ b/src/vmm/src/gdb/target.rs
@@ -28,25 +28,11 @@ use vm_memory::{Bytes, GuestAddress};
 use super::hw_breakpoint_store::HwBpStore;
 use crate::logger::{error, info};
 use crate::vstate::vcpu::{VcpuError, VcpuSendEventError};
-use crate::{arch, FcExitCode, VcpuEvent, VcpuResponse, Vmm};
+use crate::{arch, FcExitCode, VcpuEvent, VcpuHandle, VcpuResponse, Vmm};
 
 #[cfg(target_arch = "x86_64")]
 const X86_SW_BP_OP: u8 = 0xCC;
 
-#[derive(Default, Debug, Clone)]
-/// Stores the current state of a Vcpu
-struct VcpuState {
-    single_step: bool,
-    paused: bool,
-}
-
-impl VcpuState {
-    /// Disables single stepping on the Vcpu state
-    fn reset_vcpu_state(&mut self) {
-        self.single_step = false;
-    }
-}
-
 /// Errors from interactions between GDB and the VMM
 #[derive(Debug)]
 pub enum Error {
@@ -116,13 +102,12 @@ pub struct FirecrackerTarget {
     /// Used to track the currently configured hardware breakpoints.
     /// Limited to 4 in x86 see:
     /// https://elixir.bootlin.com/linux/v6.1/source/arch/x86/include/asm/kvm_host.h#L210
-    hw_breakpoints: HwBpStore<4>,
+    pub hw_breakpoints: HwBpStore<4>,
     /// Used to track the currently configured software breakpoints and store the op-code
     /// which was swapped out
     sw_breakpoints: HashMap<<GdbArch as Arch>::Usize, u8>,
 
-    /// Stores the current state of each Vcpu
-    vcpu_state: Vec<VcpuState>,
+    nr_cpus: usize,
 
     /// Stores the current paused thread id, GDB can inact commands without providing us a Tid to
     /// run on and expects us to use the last paused thread.
@@ -158,8 +143,6 @@ impl FirecrackerTarget {
                 .len()
         };
 
-        let vcpu_state = vec![Default::default(); cpu_count];
-
         Self {
             vmm,
             entry_addr,
@@ -167,12 +150,23 @@ impl FirecrackerTarget {
             // We only support 4 hw breakpoints on x86 this will need to be configurable on arm
             hw_breakpoints: Default::default(),
             sw_breakpoints: HashMap::new(),
-            vcpu_state,
+            nr_cpus: cpu_count,
 
             paused_vcpu: Tid::new(1),
         }
     }
 
+    pub fn with_vcpu_handle<R>(&self, tid: Tid, fun: impl FnOnce(&mut VcpuHandle) -> R) -> R {
+        let mut vmm = self.vmm.lock().unwrap();
+        let vcpu_handle = &mut vmm.vcpus_handles[tid_to_vcpuid(tid)];
+
+        fun(vcpu_handle)
+    }
+
+    pub fn with_paused_handle<R>(&self, fun: impl FnOnce(&mut VcpuHandle) -> Result<R, Error>) -> Result<R, Error> {
+        self.with_vcpu_handle(self.get_paused_vcpu()?, fun)
+    }
+
     /// Retrieves the currently paused Vcpu returns an error if there is no currently paused Vcpu
     fn get_paused_vcpu(&self) -> Result<Tid, Error> {
         self.paused_vcpu.ok_or(Error::NoPausedVCpu)
@@ -181,18 +175,19 @@ impl FirecrackerTarget {
     /// Updates state to reference the currently paused Vcpu and store that the cpu is currently
     /// paused
     pub fn update_paused_vcpu(&mut self, tid: Tid) {
-        self.vcpu_state[tid_to_vcpuid(tid)].paused = true;
         self.paused_vcpu = Some(tid);
     }
 
+    pub fn active_backpoint(&self) -> Vec<GuestAddress> {
+        self.hw_breakpoints.active_breakpoint()
+    }
+
     /// Resumes execution of the vmm, this will update all paused Vcpus with current kvm debug info
     /// and resume them
     fn resume_execution(&mut self) -> Result<(), Error> {
-        for cpu_id in 0..self.vcpu_state.len() {
-            let tid = vcpuid_to_tid(cpu_id)?;
-
-            self.update_kvm_debug(tid)?;
-            self.request_resume(tid)?;
+        for handle in &mut self.vmm.lock().unwrap().vcpus_handles {
+            handle.update_kvm_debug(self.active_backpoint())?;
+            handle.request_resume()?;
         }
 
         self.paused_vcpu = None;
@@ -200,10 +195,10 @@ impl FirecrackerTarget {
         Ok(())
     }
 
-    /// Resets all Vcpus to their base state
+    /// Resets all Vcpusrequest_pause to their base state
     fn reset_all_states(&mut self) {
-        for value in self.vcpu_state.iter_mut() {
-            value.reset_vcpu_state();
+        for value in self.vmm.lock().unwrap().vcpus_handles.iter_mut() {
+            value.single_step = false;
         }
     }
 
@@ -215,38 +210,61 @@ impl FirecrackerTarget {
             .stop(FcExitCode::Ok)
     }
 
-    /// Pauses the requested Vcpu
-    pub fn request_pause(&mut self, tid: Tid) -> Result<(), Error> {
-        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
+    /// Identifies why the specifc core was paused to be returned to GDB if None is returned this
+    /// indicates to handle this internally and don't notify GDB
+    pub fn get_stop_reason(&self, tid: Tid) -> Result<Option<BaseStopReason<Tid, u64>>, Error> {
+        self.with_vcpu_handle(tid, |handle| {
+            if handle.single_step {
+                return Ok(Some(MultiThreadStopReason::SignalWithThread {
+                    tid,
+                    signal: Signal::SIGTRAP,
+                }));
+            }
 
-        if vcpu_state.paused {
-            info!("Attempted to pause a vcpu already paused.");
-            // Pausing an already paused vcpu is not considered an error case from GDB
-            return Ok(());
-        }
+            let Ok(regs) = handle.get_regs() else {
+                return Ok(Some(MultiThreadStopReason::SwBreak(tid)));
+            };
+
+            let physical_addr = handle.translate_gva(regs.rip)?;
+            if self.sw_breakpoints.contains_key(&physical_addr) {
+                return Ok(Some(MultiThreadStopReason::SwBreak(tid)));
+            }
+
+            if self.hw_breakpoints.contains(&GuestAddress(regs.rip)) {
+                return Ok(Some(MultiThreadStopReason::HwBreak(tid)));
+            }
+
+            if regs.rip == self.entry_addr.0 {
+                return Ok(Some(MultiThreadStopReason::HwBreak(tid)));
+            }
 
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+            // This is not a breakpoint we've set, likely one set by the guest
+            Ok(None)
+        })
+    }
+}
 
-        cpu_handle.send_event(VcpuEvent::Pause)?;
-        let _ = cpu_handle.response_receiver().recv()?;
+impl VcpuHandle {
+    /// Pauses the requested Vcpu
+    pub fn request_pause(&mut self) -> Result<(), Error> {
+        self.send_event(VcpuEvent::Pause)?;
+        let _ = self.response_receiver().recv()?;
 
-        vcpu_state.paused = false;
         Ok(())
     }
 
     /// Gets the register values for a Vcpu
-    fn get_regs(&self, tid: Tid) -> Result<kvm_regs, Error> {
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
-        cpu_handle.send_event(VcpuEvent::GetRegisters)?;
+    fn get_regs(&self) -> Result<kvm_regs, Error> {
+        self.send_event(VcpuEvent::GetRegisters)?;
 
-        let response = cpu_handle.response_receiver().recv()?;
+        let response = self.response_receiver().recv()?;
 
         if let VcpuResponse::KvmRegisters(response) = response {
             return Ok(response);
         }
 
         if let VcpuResponse::NotAllowed(message) = response {
-            error!("Response from get regs: {message} for TID {tid:?}");
+            error!("Response from get regs: {message} for TID ");
         }
 
         Err(Error::VcuRequestError)
@@ -254,11 +272,11 @@ impl FirecrackerTarget {
 
     /// Sets the register values for a Vcpu
     fn set_regs(&self, regs: kvm_regs, tid: Tid) -> Result<(), Error> {
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
-        cpu_handle.send_event(VcpuEvent::SetRegisters(regs))?;
+        self.send_event(VcpuEvent::SetRegisters(regs))?;
 
-        let response = cpu_handle.response_receiver().recv()?;
+        let response = self.response_receiver().recv()?;
         if let VcpuResponse::NotAllowed(message) = response {
+            // TODO: more this error message to caller
             error!("Response from set regs: {message} on tid: {tid:?}");
             return Err(Error::VcuRequestError);
         }
@@ -267,45 +285,26 @@ impl FirecrackerTarget {
     }
 
     /// Resumes the Vcpu, will return early if the Vcpu is already running
-    pub fn request_resume(&mut self, tid: Tid) -> Result<(), Error> {
-        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
-
-        if !vcpu_state.paused {
-            info!("Attempted to resume a vcpu already running.");
-            // Resuming an already running Vcpu is not considered an error case from GDB
-            return Ok(());
-        }
-
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+    pub fn request_resume(&mut self) -> Result<(), Error> {
+        self.send_event(VcpuEvent::Resume)?;
 
-        cpu_handle.send_event(VcpuEvent::Resume)?;
-
-        let response = cpu_handle.response_receiver().recv()?;
+        let response = self.response_receiver().recv()?;
         if let VcpuResponse::NotAllowed(message) = response {
             error!("Response resume : {message}");
             return Err(Error::VcuRequestError);
         }
 
-        vcpu_state.paused = false;
         Ok(())
     }
 
     /// Updates the kvm debug flags set against the Vcpu
-    fn update_kvm_debug(&self, tid: Tid) -> Result<(), Error> {
-        let vcpu_state = &self.vcpu_state[tid_to_vcpuid(tid)];
-        if !vcpu_state.paused {
-            info!("Attempted to update kvm debug on a non paused Vcpu");
-            return Ok(());
-        }
-
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
-
-        cpu_handle.send_event(VcpuEvent::SetKvmDebug(
-            self.hw_breakpoints.active_breakpoint(),
-            vcpu_state.single_step,
+    fn update_kvm_debug(&self, active_breakpoint: Vec<GuestAddress>) -> Result<(), Error> {
+        self.send_event(VcpuEvent::SetKvmDebug(
+            active_breakpoint,
+            self.single_step,
         ))?;
 
-        match cpu_handle.response_receiver().recv() {
+        match self.response_receiver().recv() {
             Ok(VcpuResponse::NotAllowed(message)) => {
                 error!("Response set kvm debug: {message}");
                 Err(Error::VcuRequestError)
@@ -317,11 +316,10 @@ impl FirecrackerTarget {
     }
 
     /// Translates from a guest virtual address to a physical address
-    fn translate_gva(&self, tid: Tid, address: u64) -> Result<u64, Error> {
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
-        cpu_handle.send_event(VcpuEvent::GvaTranslate(address))?;
+    fn translate_gva(&self,  address: u64) -> Result<u64, Error> {
+        self.send_event(VcpuEvent::GvaTranslate(address))?;
 
-        let response = cpu_handle.response_receiver().recv()?;
+        let response = self.response_receiver().recv()?;
 
         if let VcpuResponse::GvaTranslation(response) = response {
             return match response {
@@ -339,14 +337,13 @@ impl FirecrackerTarget {
     }
 
     /// Injects a software breakpoint back into the guest
-    pub fn inject_bp_to_guest(&self, tid: Tid) -> Result<(), Error> {
-        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
-        cpu_handle.send_event(VcpuEvent::InjectKvmBP(
-            self.hw_breakpoints.active_breakpoint(),
+    pub fn inject_bp_to_guest(&self, active_breakpoint: Vec<GuestAddress>) -> Result<(), Error> {
+        self.send_event(VcpuEvent::InjectKvmBP(
+            active_breakpoint,
             false,
         ))?;
 
-        match cpu_handle.response_receiver().recv() {
+        match self.response_receiver().recv() {
             Ok(VcpuResponse::NotAllowed(message)) => {
                 error!("Vcpu response to injecting bp: {message}");
                 Err(Error::VcuRequestError)
@@ -356,37 +353,6 @@ impl FirecrackerTarget {
             _ => Ok(()),
         }
     }
-
-    /// Identifies why the specifc core was paused to be returned to GDB if None is returned this
-    /// indicates to handle this internally and don't notify GDB
-    pub fn get_stop_reason(&self, tid: Tid) -> Result<Option<BaseStopReason<Tid, u64>>, Error> {
-        if self.vcpu_state[tid_to_vcpuid(tid)].single_step {
-            return Ok(Some(MultiThreadStopReason::SignalWithThread {
-                tid,
-                signal: Signal::SIGTRAP,
-            }));
-        }
-
-        let Ok(regs) = self.get_regs(tid) else {
-            return Ok(Some(MultiThreadStopReason::SwBreak(tid)));
-        };
-
-        let physical_addr = self.translate_gva(tid, regs.rip)?;
-        if self.sw_breakpoints.contains_key(&physical_addr) {
-            return Ok(Some(MultiThreadStopReason::SwBreak(tid)));
-        }
-
-        if self.hw_breakpoints.contains(&GuestAddress(regs.rip)) {
-            return Ok(Some(MultiThreadStopReason::HwBreak(tid)));
-        }
-
-        if regs.rip == self.entry_addr.0 {
-            return Ok(Some(MultiThreadStopReason::HwBreak(tid)));
-        }
-
-        // This is not a breakpoint we've set, likely one set by the guest
-        Ok(None)
-    }
 }
 
 impl Target for FirecrackerTarget {
@@ -414,7 +380,7 @@ impl Target for FirecrackerTarget {
 impl MultiThreadBase for FirecrackerTarget {
     /// Reads the registers for the Vcpu
     fn read_registers(&mut self, regs: &mut CoreRegs, tid: Tid) -> TargetResult<(), Self> {
-        let cpu_regs = self.get_regs(tid).map_err(|e| {
+        let cpu_regs = self.with_vcpu_handle(tid, |handle| handle.get_regs()).map_err(|e| {
             error!("Failed to read cpu registers: {e:?}");
             TargetError::NonFatal
         })?;
@@ -471,7 +437,7 @@ impl MultiThreadBase for FirecrackerTarget {
             rflags: regs.eflags as u64,
         };
 
-        self.set_regs(new_regs, tid).map_err(|e| {
+        self.with_vcpu_handle(tid, |handle| handle.set_regs(new_regs, tid)).map_err(|e| {
             error!("Error setting registers {e:?}");
             TargetError::NonFatal
         })
@@ -495,7 +461,7 @@ impl MultiThreadBase for FirecrackerTarget {
         while total_read < len {
             let gaddr = start_addr + total_read;
 
-            let paddr = self.translate_gva(tid, gaddr as u64).map_err(|e| {
+            let paddr = self.with_vcpu_handle(tid, |handle| handle.translate_gva(gaddr as u64)).map_err(|e| {
                 error!("Error {e:?} translating gva on read address: {start_addr:X}");
                 TargetError::NonFatal
             })?;
@@ -546,7 +512,7 @@ impl MultiThreadBase for FirecrackerTarget {
 
         while total_written < len {
             let gaddr = start_addr + total_written;
-            let paddr = match self.translate_gva(tid, gaddr as u64) {
+            let paddr = match self.with_vcpu_handle(tid, |handle| handle.translate_gva(gaddr as u64)) {
                 Ok(0) => gaddr,
                 Ok(paddr) => usize::try_from(paddr).map_err(|e| {
                     error!("Error converting addr to usize: {e:?}");
@@ -591,7 +557,7 @@ impl MultiThreadBase for FirecrackerTarget {
         &mut self,
         thread_is_active: &mut dyn FnMut(Tid),
     ) -> Result<(), Self::Error> {
-        for id in 0..self.vcpu_state.len() {
+        for id in 0..self.nr_cpus {
             thread_is_active(vcpuid_to_tid(id)?)
         }
 
@@ -616,8 +582,7 @@ impl MultiThreadResume for FirecrackerTarget {
         tid: Tid,
         _signal: Option<Signal>,
     ) -> Result<(), Self::Error> {
-        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
-        vcpu_state.single_step = false;
+        self.vmm.lock().unwrap().vcpus_handles[tid_to_vcpuid(tid)].single_step = false;
 
         Ok(())
     }
@@ -647,8 +612,7 @@ impl MultiThreadSingleStep for FirecrackerTarget {
         tid: Tid,
         _signal: Option<Signal>,
     ) -> Result<(), Self::Error> {
-        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
-        vcpu_state.single_step = true;
+        self.vmm.lock().unwrap().vcpus_handles[tid_to_vcpuid(tid)].single_step = true;
 
         Ok(())
     }
@@ -677,7 +641,7 @@ impl HwBreakpoint for FirecrackerTarget {
         if !self.hw_breakpoints.add(GuestAddress(addr)) {
             return Ok(false);
         }
-        self.update_kvm_debug(self.get_paused_vcpu()?)?;
+        self.with_paused_handle(|handle| handle.update_kvm_debug(self.active_backpoint()))?;
 
         Ok(true)
     }
@@ -689,7 +653,7 @@ impl HwBreakpoint for FirecrackerTarget {
         _kind: <Self::Arch as Arch>::BreakpointKind,
     ) -> TargetResult<bool, Self> {
         self.hw_breakpoints.remove(&GuestAddress(addr));
-        self.update_kvm_debug(self.get_paused_vcpu()?)?;
+        self.with_paused_handle(|handle| handle.update_kvm_debug(self.active_backpoint()))?;
 
         Ok(true)
     }
@@ -705,7 +669,7 @@ impl SwBreakpoint for FirecrackerTarget {
         addr: <Self::Arch as Arch>::Usize,
         _kind: <Self::Arch as Arch>::BreakpointKind,
     ) -> TargetResult<bool, Self> {
-        let physical_addr = self.translate_gva(self.get_paused_vcpu()?, addr)?;
+        let physical_addr = self.with_paused_handle(|handle| handle.translate_gva(addr))?;
 
         if self.sw_breakpoints.contains_key(&physical_addr) {
             return Ok(true);
@@ -728,7 +692,7 @@ impl SwBreakpoint for FirecrackerTarget {
         addr: <Self::Arch as Arch>::Usize,
         _kind: <Self::Arch as Arch>::BreakpointKind,
     ) -> TargetResult<bool, Self> {
-        let physical_addr = self.translate_gva(self.get_paused_vcpu()?, addr)?;
+        let physical_addr = self.with_paused_handle(|handle| handle.translate_gva(addr))?;
 
         if let Some(removed) = self.sw_breakpoints.remove(&physical_addr) {
             self.write_addrs(addr, &[removed], self.get_paused_vcpu()?)?;
diff --git a/src/vmm/src/vstate/vcpu/mod.rs b/src/vmm/src/vstate/vcpu/mod.rs
index 366e7555c..731d336b0 100644
--- a/src/vmm/src/vstate/vcpu/mod.rs
+++ b/src/vmm/src/vstate/vcpu/mod.rs
@@ -796,6 +796,9 @@ pub struct VcpuHandle {
     // Rust JoinHandles have to be wrapped in Option if you ever plan on 'join()'ing them.
     // We want to be able to join these threads in tests.
     vcpu_thread: Option<thread::JoinHandle<()>>,
+
+    #[cfg(feature = "debug")]
+    pub single_step: bool
 }
 
 /// Error type for [`VcpuHandle::send_event`].
@@ -819,6 +822,9 @@ impl VcpuHandle {
             event_sender,
             response_receiver,
             vcpu_thread: Some(vcpu_thread),
+
+            #[cfg(feature = "debug")]
+            single_step: false
         }
     }
     /// Sends event to vCPU.

```

</details>